File: src/audit/audit.ts
// Audit the resolved dependency set against an AuditSource. Print-free: builds
// an AuditReport the CLI renders.

import { type Coordinates, coordinatesToString } from "../packages/index.ts";
import {
  type Advisory,
  type AuditReport,
  type AuditSource,
  type PackageAdvisories,
  type Severity,
  SEVERITY_ORDER,
} from "./types.ts";

const rank = (s: Severity): number => SEVERITY_ORDER.indexOf(s);

function worstSeverity(advisories: readonly Advisory[]): Severity {
  return advisories.reduce<Severity>(
    (worst, a) => (rank(a.severity) < rank(worst) ? a.severity : worst),
    "unknown",
  );
}

/** Scan `coordinates` (deduped) for known vulnerabilities. */
export async function auditPackages(
  coordinates: readonly Coordinates[],
  source: AuditSource,
): Promise<AuditReport> {
  // dedupe by exact coordinate (a package can appear in several lock sets)
  const unique = new Map<string, Coordinates>();
  for (const c of coordinates) unique.set(coordinatesToString(c), c);
  const distinct = [...unique.values()];

  const found = await source.findVulnerabilities(distinct);

  const counts: Record<Severity, number> = {
    critical: 0,
    high: 0,
    moderate: 0,
    low: 0,
    unknown: 0,
  };
  const vulnerable: PackageAdvisories[] = [];
  for (const c of distinct) {
    const advisories = found.get(coordinatesToString(c));
    if (!advisories || advisories.length === 0) continue;
    for (const a of advisories) counts[a.severity]++;
    vulnerable.push({ coordinates: c, advisories });
  }

  // worst severity first, then by coordinate for stable output
  vulnerable.sort((a, b) => {
    const byWorst = rank(worstSeverity(a.advisories)) - rank(worstSeverity(b.advisories));
    return byWorst !== 0
      ? byWorst
      : coordinatesToString(a.coordinates).localeCompare(coordinatesToString(b.coordinates));
  });

  return { scanned: distinct.length, vulnerable, counts };
}

File: togo/internal/audit/audit.go
package audit

import (
	"cmp"
	"slices"

	"github.com/nikeee/cappu/internal/packages"
)

// worstSeverity is the highest-ranked severity among advisories.
func worstSeverity(advisories []Advisory) Severity {
	worst := SeverityUnknown
	for _, a := range advisories {
		if severityRank(a.Severity) < severityRank(worst) {
			worst = a.Severity
		}
	}
	return worst
}

// AuditPackages scans coordinates (deduped) for known vulnerabilities. Port of
// auditPackages.
func AuditPackages(coordinates []packages.Coordinates, source AuditSource) (AuditReport, error) {
	// dedupe by exact coordinate (a package can appear in several lock sets),
	// preserving first-seen order
	seen := map[packages.CoordinateString]struct{}{}
	var distinct []packages.Coordinates
	for _, c := range coordinates {
		key := c.String()
		if _, ok := seen[key]; ok {
			continue
		}
		seen[key] = struct{}{}
		distinct = append(distinct, c)
	}

	found, err := source.FindVulnerabilities(distinct)
	if err != nil {
		return AuditReport{}, err
	}

	report := AuditReport{Scanned: len(distinct)}
	for _, c := range distinct {
		advisories := found[c.String()]
		if len(advisories) == 0 {
			continue
		}
		for _, a := range advisories {
			report.Counts.inc(a.Severity)
		}
		report.Vulnerable = append(report.Vulnerable, PackageAdvisories{Coordinates: c, Advisories: advisories})
	}

	// worst severity first, then by coordinate for stable output
	slices.SortStableFunc(report.Vulnerable, func(a, b PackageAdvisories) int {
		if byWorst := severityRank(worstSeverity(a.Advisories)) - severityRank(worstSeverity(b.Advisories)); byWorst != 0 {
			return byWorst
		}
		return cmp.Compare(a.Coordinates.String(), b.Coordinates.String())
	})
	return report, nil
}
